packages/babel-plugin-transform-block-scoping/src/index.js

```diff
@@ -545,7 +544,9 @@ class BlockScoping {
     this.hoistVarDeclarations();
 
     // turn outsideLetReferences into an array
-    const args = values(outsideRefs).map(id => t.cloneNode(id));
```

Comment:
Note that `Object.keys()` will throw on nullish values but we are good here because it is initialized as `Object.create(null)`.

I think an alternative way is refactoring `outsideRefs` as a `Map()`, so we will have `.keys()` and `.values()`, and what's more the `outsideRefs`, by design, does not have fixed shape, so V8 will finally gave up due to [megamorphic shapes](https://erdem.pl/2019/08/v-8-function-optimization#inline-cache-ic).

Reply:
Thanks @JLHwung - just to confirm, that would include refactoring the variable referenced by `this.outsideLetReferences`, as well as the local variable `outsideRefs`?

Reply:
Is there a reason we can’t use `Object.values().map` here?

Reply:
@kaicataldo We should be set for `Object.values` from Babel 8 onwards; but `main`'s using Node [6.x](https://github.com/babel/babel/blob/3e55270135410c5b0f75de7b9a283565357fe595/package.json#L82) and `Object.values` [isn't available there](https://node.green/#ES2017-features-Object-static-methods-Object-values) currently.

Reply:
@jayaddison Yep it may end up with more efforts so it's not a blocker and more like good-to-have.

Reply:
Ok; it does look like a little more work, but not too much, and it could be a valuable cleanup (and performance improvement if I understand the V8 implications correctly).  I'll give that a try here.

Reply:
(NB: unfortunately I can't easily run tests locally at the moment, so this may involve a few attempts.  I'll spend more time getting those to work if this ends up with repeat CI failures..)

Reply:
@JLHwung Based on some performance regression we seem to be hitting when using `Map`, that approach is looking a little more challenging at the moment.  If we have any more ideas for how to speed things up (or if I've introduced any mistakes here that have led to the performance issues) it might still be possible.